web/src/main/scala/quasar/api/services/data.scala

```diff
@@ -0,0 +1,114 @@
+
+  type FSTask[A] = FileSystemErrT[Task, A]
+
+  /** Flatten by inserting the [[quasar.fs.FileSystemError]] into the failure case of the [[scalaz.concurrent.Task]] */
+  val flatten = new (FSTask ~> Task) {
+    def apply[A](t: FSTask[A]): Task[A] =
+      t.fold(e => Task.fail(new RuntimeException(e.shows)), Task.now).join
+  }
+
+  def service[S[_]: Functor](f: S ~> Task)(implicit R: ReadFile.Ops[S],
+                                                    W: WriteFile.Ops[S],
+                                                    M: ManageFile.Ops[S]): HttpService = {
+
+    HttpService {
+      case req @ GET -> AsPathyPath(path) :? Offset(offsetParam) +& Limit(limitParam) => {
+        val offsetWithDefault = offsetParam.getOrElse(Natural._0.successNel)
+        val offsetWithErrorMsg: String \/ Natural = offsetWithDefault.disjunction.leftMap(
+          nel => s"invalid offset: ${nel.head.sanitized} (${nel.head.details})")
+        val limitWithErrorMsg: String \/ Option[Positive] = limitParam.traverseU(_.disjunction.leftMap(
+          nel => s"invalid limit: ${nel.head.sanitized} (${nel.head.details})"))
+        val possibleResponse = for {
+          offset <- offsetWithErrorMsg
+          limit <- limitWithErrorMsg
+        } yield {
+          val requestedFormat = MessageFormat.fromAccept(req.headers.get(Accept))
+          def formatAsHttpResponse[A: EntityEncoder](data: Process[R.M, A], contentType: `Content-Type`): Task[Response] = {
+            val trans: R.F ~> Task = hoistFree(f)
+            val trans2: R.M ~> FSTask  = Hoist[FileSystemErrT].hoist(trans)
+            def withFormatHeaders(resp: Response) = {
+              val headers = contentType :: requestedFormat.disposition.toList
+              resp.putHeaders(headers: _*)
+            }
+            data.translate(trans2).unconsOption.fold(
+              fileSystemErrorResponse,
+              _.fold(Ok(""))({ case (first, rest) =>
+                Ok(Process.emit(first) ++ rest.translate(flatten))
+              }).map(withFormatHeaders)).join
+          }
+          path.fold(
+            dirPath => formatAsHttpResponse(
+              data = zippedBytes[S](dirPath, requestedFormat, offset, limit),
+              contentType = `Content-Type`(MediaType.`application/zip`)
+            ),
+            filePath => formatAsHttpResponse(
+              data = requestedFormat.encode(R.scan(filePath, offset, limit)),
+              contentType = `Content-Type`(requestedFormat.mediaType, Some(Charset.`UTF-8`))
+            )
+          )
+        }
+        possibleResponse.leftMap(errMessage => BadRequest(errMessage)).merge
+      }
+    }
+  }
+
+  def zippedBytes[S[_]: Functor](dir: AbsDir[Sandboxed], format: MessageFormat, offset: Natural, limit: Option[Positive])
+                 (implicit R: ReadFile.Ops[S], M: ManageFile.Ops[S]): Process[R.M, ByteVector] = {
+    Process.await(M.descendantFiles(dir)) { files =>
```

Comment:
Hm, maybe `descendantFiles` deserves a refactor into `Process` or at least a version that exposes the `StreamT` so that we can preserve streaming for applications like this one that can benefit from it.